tracing/src/main/java/com/palantir/tracing/Span.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ */
+
+package com.palantir.tracing;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.common.base.Optional;
+import org.immutables.value.Value;
+
+/**
+ * A value class representing a completed Span.
+ */
+@JsonDeserialize(as = ImmutableSpan.class)
+@JsonSerialize(as = ImmutableSpan.class)
+@Value.Immutable
+@Value.Style(visibility = Value.Style.ImplementationVisibility.PACKAGE)
+public abstract class Span {
+
+    public abstract String getTraceId();
+
+    public abstract Optional<String> getParentSpanId();
+
+    public abstract String getSpanId();
+
+    public abstract String getOperation();
+
+    public abstract long getStartTimeMs();
+
+    public abstract long getDurationNs();
```

Comment:
ok, that sounds fine

On Wed, May 18, 2016 at 6:35 PM Mark Elliot <notifications@github.com>
wrote:

> In tracing/src/main/java/com/palantir/tracing/Span.java
> <https://github.com/palantir/http-remoting/pull/115#discussion_r63811121>:
>
> > +@JsonSerialize(as = ImmutableSpan.class)
> > +@Value.Immutable
> > +@Value.Style(visibility = Value.Style.ImplementationVisibility.PACKAGE)
> > +public abstract class Span {
> > +
> > +    public abstract String getTraceId();
> > +
> > +    public abstract Optional<String> getParentSpanId();
> > +
> > +    public abstract String getSpanId();
> > +
> > +    public abstract String getOperation();
> > +
> > +    public abstract long getStartTimeMs();
> > +
> > +    public abstract long getDurationNs();
>
> Accurate timing on the JVM needs to use the nanosecond counter on the CPU,
> but wall clock time is only really available in milliseconds. I'm happy to
> convert the nanoseconds to milliseconds but I figured I'd save the math
> operation and leave that to a subscriber. (Don't have a strong opinion,
> just explaining rationale.)
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/palantir/http-remoting/pull/115/files/cef3b7359acf8d47d417137799a5f5547358e14e#r63811121>
>